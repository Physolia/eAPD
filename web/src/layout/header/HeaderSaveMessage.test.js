import { shallow } from 'enzyme';
import React from 'react';

import {
  plain as HeaderSaveMessage,
  mapStateToProps
} from './HeaderSaveMessage';

describe('Header save message component', () => {
  it('renders when actively saving', () => {
    expect(
      shallow(
        <HeaderSaveMessage
          isSaving
          error={false}
          lastSaved="2020-01-01T17:00:00.000Z"
        />
      )
    ).toMatchSnapshot();
  });

  it('renders correctly when an error saving occurs', () => {
    expect(
      shallow(
        <HeaderSaveMessage
          isSaving={false}
          error
          lastSaved="2020-01-01T17:00:00.000Z"
        />
      )
    ).toMatchSnapshot();
  });

  it('renders when not currently saving', () => {
    expect(
      shallow(
        <HeaderSaveMessage
          isSaving={false}
          error={false}
          lastSaved="2020-01-01T17:00:00.000Z"
        />
      )
    ).toMatchSnapshot();
  });

  it('waits a minimum time to visually switch from saving to not saving', () => {
    jest.useFakeTimers();

    const component = shallow(
      <HeaderSaveMessage
        isSaving={false}
        error={false}
        lastSaved="2020-01-01T17:00:00.000Z"
      />
    );

    // Expect to begin in a "not saving" state
    expect(component).toMatchSnapshot();

    component.setProps({ isSaving: true });
    component.update();

    // Should immediately switch to the "saving" state
    expect(component).toMatchSnapshot();

    component.setProps({
      isSaving: false,
      lastSaved: '2020-02-02T14:00:00.000Z'
    });
    component.update();

    // Should still be in the "saving" state. It should remain in this state
    // until the  750ms delay has passed.
    expect(component).toMatchSnapshot();

    jest.advanceTimersByTime(700);
    // Should still be in the "saving" state.
    expect(component).toMatchSnapshot();

    jest.advanceTimersByTime(50);
    // Should now have switched back to the "saved" state.
    expect(component).toMatchSnapshot();

    jest.useRealTimers()
  });

  it('maps redux state to component props', () => {
    const state = {
      saving: {
        lastSaved: 'blorp',
        saving: 'bloop',
        error: false
      }
    };

    expect(mapStateToProps(state)).toEqual({
      isSaving: 'bloop',
      lastSaved: 'blorp',
      error: false
    });
  });
});
