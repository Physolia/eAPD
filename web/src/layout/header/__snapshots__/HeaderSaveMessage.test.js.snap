// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header save message component renders correctly when an error saving occurs 1`] = `
<span>
  <Check />
   
  <SaveMessage
    error={true}
    lastSaved="2020-01-01T17:00:00.000Z"
  />
</span>
`;

exports[`Header save message component renders when actively saving 1`] = `
<span>
  <Spinner
    spin={true}
  />
   Saving...
</span>
`;

exports[`Header save message component renders when not currently saving 1`] = `
<span>
  <Check />
   
  <SaveMessage
    error={false}
    lastSaved="2020-01-01T17:00:00.000Z"
  />
</span>
`;

exports[`Header save message component waits a minimum time to visually switch from saving to not saving 1`] = `
<span>
  <Check />
   
  <SaveMessage
    error={false}
    lastSaved="2020-01-01T17:00:00.000Z"
  />
</span>
`;

exports[`Header save message component waits a minimum time to visually switch from saving to not saving 2`] = `
<span>
  <Spinner
    spin={true}
  />
   Saving...
</span>
`;

exports[`Header save message component waits a minimum time to visually switch from saving to not saving 3`] = `
<span>
  <Spinner
    spin={true}
  />
   Saving...
</span>
`;

exports[`Header save message component waits a minimum time to visually switch from saving to not saving 4`] = `
<span>
  <Spinner
    spin={true}
  />
   Saving...
</span>
`;

exports[`Header save message component waits a minimum time to visually switch from saving to not saving 5`] = `
<span>
  <Check />
   
  <SaveMessage
    error={false}
    lastSaved="2020-02-02T14:00:00.000Z"
  />
</span>
`;
